Helper: First pass at merging the Helper plugin into WooCommerce

kovshenin opened the pull request:
To keep non WordPress.org-hosted extensions up to date, WooCommerce currently uses a separate plugin called WooCommerce Helper. We're working towards bringing all the functionality that plugin provides (and more!) into WooCommerce Core. This first pass allows users to: 

* Connect their store to a WooCommerce.com account
* List their available product subscriptions
* Activate/deactivate subscriptions and extensions
* Receive package updates for active subscriptions
* Receive update notifications for inactive/unavailable subscriptions

If you experience any trouble while testing this, please turn on `WP_DEBUG` and look at the `helper` log in System Status for more insight.

kovshenin committed:
Helper: First pass at merging the Helper plugin into WooCommerce

* Connect a site to a WooCommerce.com account
* List available product subscriptions
* Activate/deactivate subscriptions and extensions
* Serve updates for active subscriptions

kovshenin committed:
Helper: Fix some phpcs issues

mikejolley commented on assets/css/helper.scss:
```diff
@@ -0,0 +1,160 @@
+		background-color: #ffffff;
+	}
+
+	table.widefat, .wp-editor-container, .stuffbox, p.popular-tags, .widgets-holder-wrap, .popular-tags, .feature-filter, .imgedit-group, .comment-ays {
```
Place each selector on a new line (https://make.wordpress.org/core/handbook/best-practices/coding-standards/css/#structure).

mikejolley commented on includes/admin/helper/class-wc-helper-api.php:
```diff
@@ -0,0 +1,112 @@
+		}
+
+		// TODO: Check response signatures on certain endpoints.
+		return wp_remote_request( $url, $args );
```
Question: Default timeout is 3 secs. Is this sufficient?

mikejolley replied:
`wp_safe_remote_request` should be used here? `woocommerce_helper_api_base` is filterable.

kovshenin replied:
Core's default timeout is 5 seconds. Unless WooCommerce changes this with `http_request_timeout` then it should be sufficient.

kovshenin replied:
b4ba43f551fe2fb25774455431548911c8637426

mikejolley commented on includes/admin/helper/class-wc-helper-plugin-info.php:
```diff
@@ -0,0 +1,60 @@
+		// Look through local Woo plugins by slugs.
+		foreach ( WC_Helper::get_local_woo_plugins() as $plugin ) {
+			$slug = dirname( $plugin['_filename'] );
+			if ( dirname( $plugin['_filename'] ) == $args->slug ) {
```
===

mikejolley commented on includes/admin/helper/views/html-main.php:
```diff
@@ -0,0 +1,188 @@
+	<?php include( WC_Helper::get_view_filename( 'html-section-notices.php' ) ); ?>
+	<?php include( WC_Helper::get_view_filename( 'html-section-account.php' ) ); ?>
+
+	<h2>Subscriptions</h2>
```
This file is missing localisation on all many of the strings - flagging this one but there are more below.

mikejolley commented on includes/admin/helper/views/html-main.php:
```diff
@@ -0,0 +1,188 @@
+						<?php endif; ?>
+
+						<?php if ( ! $installed ) : ?>
+							<strong><?php echo $updates[ $product_id ]['version']; ?> available</a></strong>
```
Missing escape

mikejolley commented on includes/admin/helper/views/html-main.php:
```diff
@@ -0,0 +1,188 @@
+							<strong><?php echo date( 'F jS, Y', $subscription['expires'] ); ?></strong>
+						<?php else : ?>
+							<span class="renews" style="color: black;">Expires on:</span><br>
+							<span><?php echo date( 'F jS, Y', $subscription['expires'] ); ?></span>
```
`date_i18n` here and lines 84, 81, 78

mikejolley commented on includes/admin/helper/views/html-oauth-start.php:
```diff
@@ -0,0 +1,20 @@
+			<div class="text">
+				<img src="https://woocommerce.com/wp-content/themes/woomattic/images/logo-woocommerce@2x.png" alt="WooCommerce" style="width:180px;">
+
+				<?php if ( ! empty( $_GET['wc-helper-status'] ) && $_GET['wc-helper-status'] = 'helper-disconnected' ) : ?>
```
`$_GET['wc-helper-status'] = 'helper-disconnected'`

you're setting GET to `'helper-disconnected'` here.

kovshenin replied:
Good catch :)

mikejolley commented on includes/admin/helper/class-wc-helper-updater.php:
```diff
@@ -0,0 +1,218 @@
+	exit;
+}
+
+class WC_Helper_Updater {
```
Just a nitpick; some classes and methods throughout the PR are missing docblocks.

mikejolley reviewed (changes requested):
Given the code a quick scan and pointed out some potential issues. Leaving additional comments in a sec.

mikejolley commented:
Had a quick test of this too @kovshenin and functionality worked flawlessly. I have a few expired products on my account which showed fine, as did the plugins I have installed which have no subs.

Just some ideas (forgive me if this has been discussed in other places):

![2017-04-24 at 14 47](https://cloud.githubusercontent.com/assets/90977/25340326/47ccbbee-28fd-11e7-823d-d3989c51ba03.png)

1. Helper is currently 'always on'. Should we a) add a way to disable this (some users have moaned about the extensions screen before and requested a filter to disable), and b) does it have to always be included, or can be do some kind of detection to see if it's needed or not? i.e. Woo plugin installed.

2. We have 2 sub-menus; extensions and helper. I don't recall where but I do remember a discussion about users expecting 'extensions' to be related to installed extensions. Why can't we merge these sections? Helper would fit quite nicely under there as a sub-menu item or like this:

![2017-04-24 at 14 54](https://cloud.githubusercontent.com/assets/90977/25340527/f13943e6-28fd-11e7-9f3a-3c44258f1541.png)

The connect banner shown on the helper screen (maybe a more compact version) could also go in there when needed?

cc @mattyza 

3. The actual menu item could show a count or alert symbol if action is required, like the 'plugins' menu does for updates.

claudiosanches commented on includes/admin/helper/views/html-oauth-start.php:
```diff
@@ -0,0 +1,20 @@
+				<img src="https://woocommerce.com/wp-content/themes/woomattic/images/logo-woocommerce@2x.png" alt="WooCommerce" style="width:180px;">
+
+				<?php if ( ! empty( $_GET['wc-helper-status'] ) && $_GET['wc-helper-status'] = 'helper-disconnected' ) : ?>
+					<p><strong>Sorry to see you go</strong>. Feel free to reconnect again using the button below.</p>
```
This need to be hard coded?

claudiosanches commented on includes/admin/helper/views/html-oauth-start.php:
```diff
@@ -0,0 +1,20 @@
+
+				<h2><?php _e( 'Connect your store to your WooCommerce.com account, and manage your purchases directly from your personal site', 'woocommerce' ); ?></h2>
+				<p><?php _e( 'When you connect your account, your purchases will automatically be listed on your site. You can disconnect the account at anytime. Once your connected, you will receive the updates &amp; support that come with your subscription. ', 'woocommerce' ); ?></p>
+				<p><a class="button-primary" href="<?php echo esc_url( $connect_url ); ?>">Connect your store</a></p>
```
Another hard coded string.

claudiosanches commented on includes/admin/helper/views/html-section-account.php:
```diff
@@ -0,0 +1,16 @@
+<?php defined( 'ABSPATH' ) or exit(); ?>
+
+<h2>Your Account</h2>
```
Other hard coded string.

claudiosanches commented:
There is a few missing docblocks and hardcoded strings that requires to be fixed.

mattyza commented:
+1 to merging the "Extensions" and "Helper" screens. I believe this is/was always the intention, going in.

To the point about selectively disabling the Helper portion of the screen, perhaps we could do this automagically, through detection of the products using that screen? If no extensions with product keys are detected, don't show the screen. If extensions with product keys are present, we then show the screen.

Thoughts @mikejolley @kovshenin @garymurray ?

garymurray commented:
> merging the "Extensions" and "Helper" screens. I believe this is/was always the intention, going in.

Yes, I am going to look at how best we can merge those two screens together in some way. Its next on my agenda.

kovshenin committed:
Helper: Use `wp_safe_remote_request` for external calls.

Also whitelists woocommerce.dev as an external URL if
WP_DEBUG is turned on.

kovshenin committed:
Helper: Fix CSS selectors structure

See: https://make.wordpress.org/core/handbook/best-practices/coding-standards/css/#structure

kovshenin committed:
Helper: Strict equality check on slugs in `plugin_information` calls

kovshenin committed:
Helper: Yoda condition in _http_request_host_is_external

kovshenin committed:
Helper: Add l10n functions to all strings

kovshenin committed:
Helper: Use date_i18n() instead of date()

kovshenin committed:
Helper: Don't use assignment in helper disconnect check

kovshenin committed:
Helper: Add some missing docblocks

kovshenin commented:
Thanks so much for the reviews @claudiosanches and @mikejolley! I think I've addressed all of the issues in the latest chunk of commits.

I also like the idea of merging it with the Extensions screen, we talked about it a bit with @garymurray who's working on some mockups/flows, which will likely require some changes in the views, as well as the "routing" logic in the `WC_Helper` class.

claudiosanches commented on includes/admin/helper/class-wc-helper-api.php:
```diff
@@ -0,0 +1,135 @@
+	 * @param string $url The request URI.
+	 * @param array $args By-ref, the args that will be passed to wp_remote_request().
+	 *
+	 * @return null
```
This function don't return anything, can just remove it.

claudiosanches commented on includes/admin/helper/class-wc-helper-api.php:
```diff
@@ -0,0 +1,135 @@
+
+	/**
+	 * Wrapper for self::request().
+	 */
```
Missing docblocks for `$endpoint`, `$args` and returned value.

claudiosanches commented on includes/admin/helper/class-wc-helper-api.php:
```diff
@@ -0,0 +1,135 @@
+
+	/**
+	 * Wrapper for self::request().
+	 */
```
Missing docblocks for `$endpoint`, `$args` and returned value.

claudiosanches commented on includes/admin/helper/class-wc-helper-options.php:
```diff
@@ -0,0 +1,49 @@
+ * An interface to the woocommerce_helper_data entry in the wp_options table.
+ */
+class WC_Helper_Options {
+	private static $option_name = 'woocommerce_helper_data';
```
Missing docblock.

claudiosanches commented on includes/admin/helper/class-wc-helper-api.php:
```diff
@@ -0,0 +1,135 @@
+	 * @return bool True if the host is external, false if local/internal.
+	 */
+	public static function _http_request_host_is_external( $r, $host ) {
+		if ( defined( 'WP_DEBUG' ) && WP_DEBUG && 'woocommerce.dev' === $host ) {
```
This method looks for internal tests only... Not sure if we should include it for all users.

claudiosanches commented on includes/admin/helper/class-wc-helper-api.php:
```diff
@@ -0,0 +1,135 @@
+		}
+
+		// Disable SSL verification for the dev environment.
+		if ( defined( 'WP_DEBUG' ) && WP_DEBUG && strpos( self::$api_base, 'https://woocommerce.dev/' ) === 0 ) {
```
Not sure if we should do it for all users, if we are testing it internally we could create a plugin to do it.

claudiosanches commented on includes/admin/helper/class-wc-helper-updater.php:
```diff
@@ -0,0 +1,240 @@
+			'authenticated' => true,
+		) );
+
+		if ( wp_remote_retrieve_response_code( $request ) != 200 ) {
```
Needs strict comparison

kovshenin committed:
Helper: Fix some missing docblocks

kovshenin committed:
Helper: Use strict comparison when checking response code from update-check

kovshenin committed:
Helper: Remove everything related to woocommerce.dev

kovshenin committed:
Helper: Toggle subscription on plugin activation/deactivation

When a user activates or deactivates an extension, attempt to find
a valid subscription for that extensioni, and activate/deactivate it
to spare the user an extra step.

kovshenin committed:
Helper: Fix duplicate product output

kovshenin committed:
Helper: Add contextual nags about available extension updates

kovshenin committed:
Helper: Remove some legacy updater notices

kovshenin committed:
Helper: Add better update URLs to plugin subscriptions

kovshenin committed:
Helper: Add better update URLs to theme subscriptions

kovshenin committed:
Helper: Avoid an undefined warning in the main template

kovshenin committed:
Helper: Add tabbed interface to WooCommerce > Extensions

kovshenin committed:
Helper: Redesign (work in progress)

kovshenin committed:
Helper: Clean up navigation/headings, remove some noop markup.
